rust-runtime/aws-smithy-runtime-api/src/shared.rs

```diff
@@ -0,0 +1,137 @@
+}
+
+/// Given a `value`, determine if that value is already shared. If it is, return it. Otherwise, wrap it in a shared type.
+pub fn maybe_shared<Shared, MaybeShared, F>(value: MaybeShared, ctor: F) -> Shared
+where
+    Shared: 'static,
+    MaybeShared: IntoShared<Shared> + 'static,
+    F: FnOnce(MaybeShared) -> Shared,
+{
+    // Check if the type is already a shared type
+    if TypeId::of::<MaybeShared>() == TypeId::of::<Shared>() {
+        // Convince the compiler it is already a shared type and return it
+        let mut placeholder = Some(value);
+        let value: Shared = (&mut placeholder as &mut dyn Any)
+            .downcast_mut::<Option<Shared>>()
+            .expect("type checked above")
+            .take()
+            .expect("set to Some above");
+        value
+    } else {
+        (ctor)(value)
+    }
+}
+
+/// Implements `FromUnshared` for a shared type.
+///
+/// # Example
+/// ```rust,no_run
+/// use aws_smithy_runtime_api::impl_shared_conversions;
+/// use std::sync::Arc;
+///
+/// trait Thing {}
+///
+/// struct Thingamajig;
+/// impl Thing for Thingamajig {}
+///
+/// struct SharedThing(Arc<dyn Thing>);
+/// impl Thing for SharedThing {}
+/// impl SharedThing {
+///     fn new(thing: impl Thing + 'static) -> Self {
+///         Self(Arc::new(thing))
+///     }
+/// }
+/// impl_shared_conversions!(convert SharedThing from Thing using SharedThing::new);
+/// ```
+#[macro_export]
+macro_rules! impl_shared_conversions {
+    (convert $shared_type:ident from $unshared_trait:ident using $ctor:expr) => {
+        impl<T> $crate::shared::FromUnshared<T> for $shared_type
+        where
+            T: $unshared_trait + 'static,
+        {
+            fn from_unshared(value: T) -> Self {
+                $crate::shared::maybe_shared(value, $ctor)
+            }
+        }
+    };
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fmt;
+    use std::sync::Arc;
+
+    trait Thing: fmt::Debug {}
+
+    #[derive(Debug)]
+    struct Thingamajig;
+    impl Thing for Thingamajig {}
+
+    #[derive(Debug)]
+    struct SharedThing(Arc<dyn Thing>);
+    impl Thing for SharedThing {}
+    impl SharedThing {
+        fn new(thing: impl Thing + 'static) -> Self {
+            Self(Arc::new(thing))
+        }
+    }
+    impl_shared_conversions!(convert SharedThing from Thing using SharedThing::new);
+
+    #[test]
+    fn test() {
+        let thing = Thingamajig;
+        assert_eq!("Thingamajig", format!("{thing:?}"), "precondition");
+
+        let shared_thing: SharedThing = thing.into_shared();
+        assert_eq!(
+            "SharedThing(Thingamajig)",
+            format!("{shared_thing:?}"),
+            "precondition"
+        );
+
+        let very_shared_thing: SharedThing = shared_thing.into_shared();
+        assert_eq!(
+            "SharedThing(Thingamajig)",
+            format!("{very_shared_thing:?}"),
+            "it should not nest the shared thing in another shared thing"
+        );
+    }
```

Comment:
should we assert via downcasting that an already shared type is not double boxed?

Reply:
This seems to be very non-trivial.